### ISSUE TITLE
New I/O wait model

### ISSUE THREAD

**Issue by zonyitoo** [zonyitoo/coio-rs#37]
## Motivation

Current implementation

1. Requires communications between worker threads and I/O thread for every I/O wait operation, which results in a significant communication overhead between threads.
2. Relies on `PollOpt::oneshot()` behavior, which means that it requires to `register()` and `deregister()` the I/O object on the `EventLoop` each time an event is being awaited.

## Proposal

0. Introduce a `type State = [CoroMonoBarrier; 4]`.

    * This type will be shared between the `Handler` and I/O objects and will store parked coroutines and the state of events at the same time. Each of the 4 `CoroMonoBarrier`s will correspond to one of the 4 mio event types: `readable`, `writeable`, `hup`, and `error`.

1. When the I/O object whishes to do an I/O operation:

    1. Check if it hasn't been registered with the `EventLoop` yet. If that's the case it registers itself using it's internal mio I/O object and a new `Arc<State>`, which will be shared between the `Handler` and the I/O object. This could be done using the `EventLoop::channel()`. The coroutine should be blocked until the registration has been confirmed.
> This lazy registration allows us in the case of TCP sockets to delay the channel operation until the socket is first being used inside it's newly spawned coroutine. If this would be done on the TcpListener coroutine instead it would reduce the amount of `accept()`s per second.

    2. Read or write using logic similiar to the following pseudo-code (here: reading from socket):
```rust
loop {
    // wait until we snatch READY from the barrier
    self.barriers.readable.wait();

    // try to read from the socket
    match self.try_read(...) {
        Err(err) => {
            // a failure but it's not WouldBlock! ---> put the READY flag back in
            self.barriers.readable.signal();
            return Err(err);
        }
        Ok(len) => {
            // success ---> put the READY flag back in
            self.barriers.readable.signal();
            return Ok(len);
        }
        Err(WouldBlock) => {
            // we stole READY ---> loop again
            // (the wait() above will block until it's made READY by the Handler)
        }
    }
}
```

2. When a particular event happens, `EventLoop` will `notify()` the corresponding `CoroMonoBarrier` in the `Arc<State>` to either signify a READY state or to wake up a possibly blocked Coroutine.

3. When a I/O object is going to be dropped:

    1. Check if the I/O object is registered with the `EventLoop` and `deregister()` it in that case. The `EventLoop::channel()` could be used for this.

### I/O Object

For each I/O object we need to store:

* An `Arc<State>` which shares state between the object and the `Handler`
* A `Token` to `deregister()` from the `EventLoop` again

### EventLoop

* Register all I/O event with `PollOpt::edge()`
* *TO BE DISCUSSED*: Use two separated `EventLoop`s for read events and write events

## Implementation Detail

*TO BE DISCUSSED*

**Comment by lhecker** [comment]
We should keep the first comment (yours) up to date with every new idea or correction we have, @zonyitoo.

First of all 1.2) is bad because `reregister()` can't be done on the worker thread (not supported by epoll), thus requires the usage of `EventLoop::handle()`. Furthermore it results in 1 syscall on every major platform.
This would thus only be slightly better than our current implementation, which also results in the usage of the `EventLoop::handle()` and 2 syscalls (for `register()` and `deregister()`).

What I originally proposed was that we register IO objects (sockets) for _every_ event type (i.e.: with `EventSet::all()`) and that we use an `Arc<[CoroMonoBarrier; 4]>`. Each of the 4 barriers would represent one of the 4 event types: `[CoroMonoBarrier; 4]` would basically represent `["readable", "writeable", "hup", "error"]` if we make the position in the array the same as [`log2(event_set.bits())`](https://github.com/carllerche/mio/blob/6d8f2e614c59c1b2ff2700912b696d6ba8735896/src/event.rs#L144-L161) (since we'll only support waiting for a single event type at a time only a single bit will be set).
If we want to await a particular event we will wait() on the corresponding barrier.
And whenever a event is triggered in `Handler::ready()` it should `notify()` the corresponding barrier in the Slab.

**Downside:** Possibly unnecessary calls to `Handler::ready()` will be generated by mio since we always listen for all event types. This might be okay though since we use Edge triggered events anyways.
**Upside:** We don't need to `reregister()` (= no syscalls!) and we don't need to synchronize with the `Handler` using the `EventLoop::channel()`.

What do you think about this @zonyitoo? Should I (or you) update the comment above with this?

**Comment by zonyitoo** [comment]
@lhecker This one is basically your proposal, so of course, you should modify it directly. The top of this issue is the memo for implementation.

**Comment by zonyitoo** [comment]
## Problem to be Solved

* [ ] Is it safe to call `reregister` from the worker thread?
* [ ] Does the operations discussed above could be wrapped into a single `struct IoOperation`?
* [ ] How to get the `Token` back from the `slab.insert`? Because it is required for the subsequential `reregister` and `deregister` operations.

**Comment by zonyitoo** [comment]
Implemented since 774d3f40cddf75f0024ef30ec019379926cdf0ac
